DBZ-3621 Saving Debezium offsets in JDBC database

subkanthi opened the pull request:
Added debezium-storage-jdbc. 
@jpechane , Creating a new PR as discussed in #4312. Thanks

subkanthi commented:
> @subkanthi Thanks, could you please try to address the other comment in the related PRs?

Hi @jpechane , I went through this PR https://github.com/debezium/debezium/pull/2543 and addressed the missing ones related to comments. Please let me know if you find something that is not addressed.

jpechane commented:
@subodh1810 I am sorry, probably some were lost :-(. In general I'd prefer to have at least CREAT nd SELECT queries configurable. The reasoning behind it that they might be optimized by using better dtataypes, queries on different databases.

Also could you please use naming JDBC URL, not URI? IMHO URL is definitely more often used.

jpechane commented:
@subkanthi I don't think it is necessary. It owuld be just enought to properly document each query. IMHO people will mostly change the datatypes but will keep the column names same.

jpechane commented on debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:
```diff
@@ -265,6 +265,32 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
+
+    public static final String DEFAULT_OFFSET_STORAGE_TABLE_SELECT = "SELECT id, offset_key, offset_val FROM %s " +
+            "ORDER BY record_insert_ts, record_insert_seq";
+    public static final Field OFFSET_STORAGE_TABLE_SELECT = Field.create("offset.storage.jdbc.offset_table_select`")
```
```suggestion
    public static final Field OFFSET_STORAGE_TABLE_SELECT = Field.create("offset.storage.jdbc.offset_table_select")
```

jpechane commented on debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:
```diff
@@ -265,6 +265,32 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
+     * record_insert_seq - Sequence number of record
+     */
+    public static final Field OFFSET_STORAGE_TABLE_DDL = Field.create("offset.storage.jdbc.offset_table_ddl")
+            .withDescription("Create table syntax for offset jdbc table")
```
```suggestion
            .withDescription("Create table syntax for offset JDBC table")
```

jpechane commented on debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:
```diff
@@ -265,6 +265,32 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
+    public static final String DEFAULT_OFFSET_STORAGE_TABLE_SELECT = "SELECT id, offset_key, offset_val FROM %s " +
+            "ORDER BY record_insert_ts, record_insert_seq";
+    public static final Field OFFSET_STORAGE_TABLE_SELECT = Field.create("offset.storage.jdbc.offset_table_select`")
+            .withDescription("Select syntax to get offset data from jdbc table")
```
```suggestion
            .withDescription("Select syntax to get offset data from JDBC table")
```

jpechane commented on debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:
```diff
@@ -277,7 +303,8 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
             // JDBC OFFSET STORAGE FIELDS
-            OFFSET_STORAGE_JDBC_URL, OFFSET_STORAGE_JDBC_USER, OFFSET_STORAGE_JDBC_PASSWORD, OFFSET_STORAGE_JDBC_TABLE_NAME);
+            OFFSET_STORAGE_JDBC_URL, OFFSET_STORAGE_JDBC_USER, OFFSET_STORAGE_JDBC_PASSWORD, OFFSET_STORAGE_JDBC_TABLE_NAME,
+            OFFSET_STORAGE_TABLE_DDL, OFFSET_STORAGE_TABLE_SELECT);
```
Please don't add the fields to the config. See `RedisOffsetBackingStore`

subkanthi replied:
Hi @jpechane , this was added here for the `AbstractConnectorTest`, I also checked with Chris Cranford, its duplicated unfortunately, I checked `RedisOffsetBackingStore`, I dont see an Integration test, the part I can do is to separate the config to a separate file, so its cleaner.

jpechane commented on debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:
```diff
@@ -1234,6 +1261,8 @@ protected static class EmbeddedConfig extends WorkerConfig {
             Field.group(config, "jdbc", OFFSET_STORAGE_JDBC_USER);
             Field.group(config, "jdbc", OFFSET_STORAGE_JDBC_PASSWORD);
             Field.group(config, "jdbc", OFFSET_STORAGE_JDBC_TABLE_NAME);
+            Field.group(config, "jdbc", OFFSET_STORAGE_TABLE_DDL);
```
Same as above

jpechane commented on debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConfig.java:
```diff
@@ -35,7 +37,15 @@ public class JdbcConfig extends WorkerConfig {
                         ConfigDef.Importance.HIGH,
-                        "Name of the table to store offsets");
+                        "Name of the table to store offsets")
+                .define(OFFSET_STORAGE_TABLE_DDL.name(),
```
Should not be needed here.

subkanthi replied:
WorkerConfig seems to take ConfigDef and Properties as parameters, not sure what needs to be done here.

jpechane reviewed (commented):
@subkanthi We are getting there, thanks! Could you please just revisit the config a bit more? Check the class I pointed too.

jpechane commented:
@subkanthi I rebased the PR on the latest main and made significant changes in the way how the storages are configured.
Please take a look at class `JdbcSchemaHistoryConfig` - we need to make DDL and both SELECTs configurable and then we are good to go and the PR can be merged.

subkanthi commented:
> @subkanthi I rebased the PR on the latest main and made significant changes in the way how the storages are configured. Please take a look at class `JdbcSchemaHistoryConfig` - we need to make DDL and both SELECTs configurable and then we are good to go and the PR can be merged.

Hi @jpechane , I have added the changes to make SELECT and DDL configurable for `JdbcSchemaHistoryConfig`.

jpechane commented:
@subkanthi And we are done. Thanks a lot!
